[WIN M1] WIN (ZIP/MSI) Build/Assemble Process

|Tasks(ZIP) |**Tasks(MSI)** |Estimate   |Status(ZIP)    |Status(MSI)    |Notes  |
|---    |---    |---    |---    |---    |---    |
|Re-use the existing build process to generate the  OpenSearch/Dashboards min + all of the plugins artifacts for WIN package to  use |Same   |0  |Completed  |Completed  |   |
|The artifacts should be built with ~~LINUX~~ Windows platform  specified, ~~as we will cross-compile WIN binary on LINUX then package with  WINDOWS JDK~~.  |Same   |1  |Completed    |Completed   |~~This is still in debate as we can techinically build WIN  on Windows machine, but there are a lot of things to setup just so Jenkins  can run Python on Windows Agent.~~ We are able to run shell scripts natively on Windows agent  |
|We already have "zip" supported for "—distribution“ parameter, but needs to check whether it is already combined with `--platform windows`. |We do not have "exe" support for "—distribution“ yet. However, this is different from "RPM" as we do not need min artifact to be a exe. The min artifact can be zip and the final product in assemble can be exe.  |2  |Completed   |Completed   |As for "exe" we need to discuss whether a  standard exe is enough, or do we want to invest into Windows official  installer "msi".  |
|We already have "--distribution" param available in assemble workflow, just need to verify existing functions of "ZIP".    |We already have "--distribution" param available  in assemble workflow, but no support for "EXE" redirection. Need to  add a child class supporting the new distribution.   |2  |Completed   |Completed     |   |
|The generation code should pull the artifacts from the build workflow to a temporary location  |Same   |1  |Completed   |Completed    |   |
|The code will compile the components and also call existing install function to install  plugins on min artifacts   |Same   |1  |Completed  |Completed   |  ETA: 2022/09/16 |
|After installation, the code will execute a tool or  utility to wrap all the content into corresponding distribution format |Same   |1  |Completed  |Completed    | **20220819 Note:** Plugin compilation currently have some issues with the build scripts, the compilation itself seems ok at least on things like common-utils. <br/><br/> ETA: 2022/09/16|
|The code will move the final distribution artifact from the temp location to dist folder   |Same   |1  |Completed  |Completed    |  ETA: 2022/09/07 |


### Note: MSI section in this milestone is obsolete as MSI is just a wrapper of the content in ZIP. So as long as ZIP is completed here MSI is considered complete as well.

- [ ] do not remove

* PRs:

20220715:
* https://github.com/opensearch-project/opensearch-ci/pull/155

20220721:
* https://github.com/opensearch-project/opensearch-ci/pull/167

20220722:
* https://github.com/opensearch-project/opensearch-ci/pull/169

20220819:
* https://github.com/opensearch-project/opensearch-build/pull/2483
* https://github.com/opensearch-project/opensearch-ci/pull/187

20220824:
* https://github.com/opensearch-project/opensearch-ci/pull/190

20220902:
* https://github.com/opensearch-project/opensearch-ci/pull/197
* https://github.com/opensearch-project/opensearch-build/pull/2550

20220907:
* common-utils: https://github.com/opensearch-project/common-utils/issues/238
  * https://github.com/opensearch-project/common-utils/pull/258
* https://github.com/opensearch-project/opensearch-ci/pull/199

20220914:
* alerting: https://github.com/opensearch-project/alerting/issues/557
  * https://github.com/opensearch-project/alerting/issues/573

20220915:
* https://github.com/opensearch-project/opensearch-ci/pull/201

20220916:
* https://github.com/opensearch-project/opensearch-ci/pull/202

20220927:
* https://github.com/opensearch-project/alerting/pull/570
* https://github.com/opensearch-project/alerting/pull/571
* https://github.com/opensearch-project/alerting/pull/572

20220928:
* https://github.com/opensearch-project/opensearch-build/pull/2672
* security: https://github.com/opensearch-project/security/issues/2122
* https://github.com/opensearch-project/common-utils/pull/258

20221004:
* https://github.com/opensearch-project/opensearch-build/pull/2678

20221006:
* Security: https://github.com/opensearch-project/security/issues/2148
* https://github.com/opensearch-project/opensearch-build/pull/2704

20221007:
* OSD: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2535

20221010:
* https://github.com/opensearch-project/opensearch-ci/pull/208
* https://github.com/opensearch-project/opensearch-ci/pull/209

20221011:
* https://github.com/opensearch-project/opensearch-build-libraries/pull/14
* https://github.com/opensearch-project/opensearch-build/pull/2721
* OSD: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2554

20221012:
* https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2026

20221013:
* https://github.com/opensearch-project/opensearch-build/pull/2730

20221018:
* https://github.com/opensearch-project/opensearch-build/pull/2756
* SQL: https://github.com/opensearch-project/sql/issues/928
* https://github.com/opensearch-project/opensearch-build/pull/2761

20221019:
* https://github.com/opensearch-project/OpenSearch/issues/4817

20221024:
* https://github.com/opensearch-project/opensearch-build/pull/2788

20221025:
* OSD: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2669
* https://github.com/opensearch-project/k-NN/pull/595
* https://github.com/opensearch-project/opensearch-ci/pull/212

20221027:
* https://github.com/opensearch-project/opensearch-ci/pull/213
* https://github.com/opensearch-project/opensearch-build/pull/2812
* https://github.com/opensearch-project/opensearch-build/pull/2817
* https://github.com/opensearch-project/opensearch-ci/pull/214

20221028:
* https://github.com/opensearch-project/opensearch-build/pull/2818
* https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2686

20221101:
* https://github.com/opensearch-project/opensearch-build/pull/2840
* https://github.com/opensearch-project/opensearch-build/pull/2841
* https://github.com/opensearch-project/opensearch-ci/pull/215

20221103:
* https://github.com/opensearch-project/opensearch-build/pull/2845
* https://github.com/opensearch-project/opensearch-ci/pull/219
~~* OSD: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2764~~
* OSD ftrepo: https://github.com/opensearch-project/opensearch-dashboards-functional-test/issues/370

20221104:
* https://github.com/opensearch-project/opensearch-build/pull/2848
* https://github.com/opensearch-project/opensearch-ci/pull/224
* https://github.com/opensearch-project/opensearch-build-libraries/pull/28

20221107:
* https://github.com/opensearch-project/opensearch-build/pull/2860
* https://github.com/opensearch-project/opensearch-ci/pull/228

20221108:
* OSD ftrepo: https://github.com/opensearch-project/opensearch-dashboards-functional-test/issues/370
* https://github.com/opensearch-project/opensearch-build-libraries/pull/31
* https://github.com/opensearch-project/opensearch-build/pull/2869

20221114:
* #2892

New error detected on building opensearch on Windows host:
```

C:\Users\Administrator\opensearch-build>bash ./build.sh manifests/2.1.0/opensearch-2.1.0.yml --distribution zip
Installing dependencies in . ...
Creating a virtualenv for this project...
Pipfile: C:\Users\Administrator\opensearch-build\Pipfile
Using C:/Users/Administrator/scoop/apps/python37/3.7.9/python.exe (3.7.9) to create virtualenv...
[ ===] Creating virtual environment...created virtual environment CPython3.7.9.final.0-64 in 4832ms
  creator CPython3Windows(dest=C:\Users\Administrator\.virtualenvs\opensearch-build-ppYgApDN, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=C:\Users\Administrator\AppData\Local\pypa\virtualenv)
    added seed packages: pip==22.1.2, setuptools==62.6.0, wheel==0.37.1
  activators BashActivator,BatchActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator

Successfully created virtual environment!
Virtualenv location: C:\Users\Administrator\.virtualenvs\opensearch-build-ppYgApDN
Installing dependencies from Pipfile.lock (d422cd)...
  ================================ 55/55 - 00:00:40
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
Running ./src/run_build.py manifests/2.1.0/opensearch-2.1.0.yml --distribution zip ...
2022-07-15 02:39:58 INFO     Building in C:\Users\ADMINI~1\AppData\Local\Temp\2\tmpmtqisg7j
2022-07-15 02:39:58 INFO     Building OpenSearch (x64) into C:\Users\Administrator\opensearch-build\zip\builds\opensearch
2022-07-15 02:39:58 INFO     Building OpenSearch
2022-07-15 02:39:58 INFO     Executing "git init" in C:\Users\ADMINI~1\AppData\Local\Temp\2\tmpmtqisg7j\OpenSearch
2022-07-15 02:39:58 INFO     Executing "git remote add origin https://github.com/opensearch-project/OpenSearch.git" in C:\Users\ADMINI~1\AppData\Local\Temp\2\tmpmtqisg7j\OpenSearch
2022-07-15 02:39:59 INFO     Executing "git fetch --depth 1 origin tags/2.1.0" in C:\Users\ADMINI~1\AppData\Local\Temp\2\tmpmtqisg7j\OpenSearch
2022-07-15 02:40:05 INFO     Executing "git checkout FETCH_HEAD" in C:\Users\ADMINI~1\AppData\Local\Temp\2\tmpmtqisg7j\OpenSearch
2022-07-15 02:40:11 INFO     Executing "git rev-parse HEAD" in C:\Users\ADMINI~1\AppData\Local\Temp\2\tmpmtqisg7j\OpenSearch
2022-07-15 02:40:11 INFO     Checked out https://github.com/opensearch-project/OpenSearch.git@tags/2.1.0 into C:\Users\ADMINI~1\AppData\Local\Temp\2\tmpmtqisg7j\OpenSearch at 388c80ad94529b1d9aad0a735c4740dce2932a32
2022-07-15 02:40:11 INFO     Executing "bash C:\Users\Administrator\opensearch-build\scripts\components\OpenSearch\build.sh -v 2.1.0 -p windows -a x64 -d zip -s false -o builds" in C:\Users\ADMINI~1\AppData\Local\Temp\2\tmpmtqisg7j\OpenSearch
+ getopts :h:v:q:s:o:p:a:d: arg
+ case $arg in
+ VERSION=2.1.0
+ getopts :h:v:q:s:o:p:a:d: arg
+ case $arg in
+ PLATFORM=windows
+ getopts :h:v:q:s:o:p:a:d: arg
+ case $arg in
+ ARCHITECTURE=x64
+ getopts :h:v:q:s:o:p:a:d: arg
+ case $arg in
+ DISTRIBUTION=zip
+ getopts :h:v:q:s:o:p:a:d: arg
+ case $arg in
+ SNAPSHOT=false
+ getopts :h:v:q:s:o:p:a:d: arg
+ case $arg in
+ OUTPUT=builds
+ getopts :h:v:q:s:o:p:a:d: arg
+ '[' -z 2.1.0 ']'
+ '[' -z builds ']'
+ mkdir -p builds/maven/org/opensearch
+ ./gradlew publishToMavenLocal -Dbuild.snapshot=false -Dbuild.version_qualifier=
Downloading https://services.gradle.org/distributions/gradle-7.4.2-all.zip
...............10%...............20%...............30%...............40%...............50%................60%...............70%...............80%...............90%...............100%

Welcome to Gradle 7.4.2!

Here are the highlights of this release:
 - Aggregated test and JaCoCo reports
 - Marking additional test source directories as tests in IntelliJ
 - Support for Adoptium JDKs in Java toolchains

For more details see https://docs.gradle.org/7.4.2/release-notes.html

Starting a Gradle Daemon (subsequent builds will be faster)

FAILURE: Build failed with an exception.

* What went wrong:
Unable to start the daemon process.
This problem might be caused by incorrect configuration of the daemon.
For example, an unrecognized jvm option is used.
Please refer to the User Manual chapter on the daemon at https://docs.gradle.org/7.4.2/userguide/gradle_daemon.html
Process command line: C:\Users\Administrator\scoop\apps\temurin8-jdk\8.0.332-9\bin\java.exe -XX:+HeapDumpOnOutOfMemoryError -Xss2m --add-exports jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED --add-exports jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED --add-exports jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED --add-exports jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED --add-exports jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED -Xmx3g -Dfile.encoding=windows-1252 -Duser.country=US -Duser.language=en -Duser.variant -cp C:\Users\Administrator\.gradle\wrapper\dists\gradle-7.4.2-all\9uukhhbclvbegdvsww0j0cr3p\gradle-7.4.2\lib\gradle-launcher-7.4.2.jar org.gradle.launcher.daemon.bootstrap.GradleDaemon 7.4.2
Please read the following process output to find out more:
-----------------------
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
Unrecognized option: --add-exports


* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org
2022-07-15 02:40:28 ERROR    Error building OpenSearch, retry with: ./build.sh manifests/2.1.0/opensearch-2.1.0.yml --component OpenSearch
Traceback (most recent call last):
  File "./src/run_build.py", line 81, in <module>
    sys.exit(main())
  File "./src/run_build.py", line 68, in main
    builder.build(build_recorder)
  File "C:\Users\Administrator\opensearch-build\src\build_workflow\builder_from_source.py", line 55, in build
    self.git_repo.execute(build_command)
  File "C:\Users\Administrator\opensearch-build\src\git\git_repository.py", line 83, in execute
    subprocess.check_call(command, cwd=cwd, shell=True)
  File "C:\Users\Administrator\scoop\apps\python37\3.7.9\lib\subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'bash C:\Users\Administrator\opensearch-build\scripts\components\OpenSearch\build.sh -v 2.1.0 -p windows -a x64 -d zip -s false -o builds' returned non-zero exit status 1.

C:\Users\Administrator\opensearch-build>

```

Another issue is on windows specifically after a few runs git will show malloc error:
```
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from https://github.com/peterzhuamazon/opensearch-build
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:1003)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1244)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1308)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: hudson.plugins.git.GitException: Command "git fetch --tags --force --progress -- https://github.com/peterzhuamazon/opensearch-build +refs/heads/*:refs/remotes/origin/*" returned status code 128:
stdout: 
stderr: error: Out of memory, malloc failed (tried to allocate 1132 bytes)
fatal: packed object 87d61083f41051b91815047abe51e6ff4319189f (stored in .git/objects/pack/pack-c02b411344f3128e4d33d11bce5b431cbbe6cbc8.pack) is corrupt
```

Only resolve if we set the mem limits early on:
```

git config --system core.packedGitLimit 128m
git config --system core.packedGitWindowSize 128m


git config --system pack.deltaCacheSize 128m
git config --system pack.packSizeLimit 128m
git config --system pack.windowMemory 128m
```

New issue:

```
Selected Git installation does not exist. Using Default
The recommended git tool is: NONE
No credentials specified
Cloning the remote Git repository
Cloning repository https://github.com/peterzhuamazon/opensearch-build
 > git init C:\Users\Administrator\jenkins\workspace\zhujiaxi\bundle-build-zhujiaxi # timeout=10
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Could not init C:\Users\Administrator\jenkins\workspace\zhujiaxi\bundle-build-zhujiaxi
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$5.execute(CliGitAPIImpl.java:1042)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:797)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:158)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:151)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:376)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
	Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to EC2 (Amazon_ec2_cloud) - jenkinsAgentNode-Jenkins-Agent-Windows2019-X64-C54xlarge-Single-Host (i-031d573af80effb62)
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1784)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
		at hudson.remoting.Channel.call(Channel.java:1000)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:143)
		at sun.reflect.GeneratedMethodAccessor543.invoke(Unknown Source)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:129)
		at com.sun.proxy.$Proxy85.execute(Unknown Source)
		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1226)
		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1308)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
		... 4 more
Caused by: hudson.plugins.git.GitException: Error performing git command: git init C:\Users\Administrator\jenkins\workspace\zhujiaxi\bundle-build-zhujiaxi
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2679)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2601)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2597)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1968)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$5.execute(CliGitAPIImpl.java:1040)
	... 11 more
Caused by: java.io.IOException: Cannot run program "git" (in directory "C:\Users\Administrator\jenkins\workspace\zhujiaxi\bundle-build-zhujiaxi"): CreateProcess error=2, The system cannot find the file specified
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at hudson.Proc$LocalProc.<init>(Proc.java:254)
	at hudson.Proc$LocalProc.<init>(Proc.java:223)
	at hudson.Launcher$LocalLauncher.launch(Launcher.java:997)
	at hudson.Launcher$ProcStarter.start(Launcher.java:509)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2664)
	... 15 more
Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
	at java.lang.ProcessImpl.create(Native Method)
	at java.lang.ProcessImpl.<init>(ProcessImpl.java:453)
	at java.lang.ProcessImpl.start(ProcessImpl.java:139)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 20 more
```

This is due to jenkins trying to run git without invoking windows env vars sometimes.
This seems like an issue with Jenkins Git Plugin https://issues.jenkins.io/browse/JENKINS-68580.
However, in my observation, a newly created windows agent would goes to success, after a lengthy 10min of searching for the 1st run, then succeed for about 10 more runs, and will eventually end up in this error.

In order to make sure it goes to succeed every time I will define the path of git.exe in agent, forcing Jenkins to respect the actual git location during the checkout.

This also sometimes cause broken .git folder as git commands will needs to be run multiple times for a checkout, and sometimes the 1st command succeed and the rest of the commands failed out of nowhere.

Very inconsistent, especially on windows agent.

Solution for now:
```
nodeProperties:
        - toolLocation:
            locations:
            - home: "C:\\Users\\Administrator\\scoop\\shims\\git.exe"
              key: "hudson.plugins.git.GitTool$DescriptorImpl@Default"
```

--- a/src/assemble_workflow/dist.py
+++ b/src/assemble_workflow/dist.py
@@ -103,11 +103,17 @@ class DistZip(Dist):
 
     def __build__(self, name: str, dest: str) -> None:
         with ZipFile(name, "w", zipfile.ZIP_DEFLATED) as zip:
-            rootlen = len(self.archive_path) + 1
+            # root               : /tmp/tmp********/opensearch-<version+qualifier>
+            # leadingdir         : opensearch-<version+qualifier>
+            # root no leading dir: /tmp/tmp********/
+            # This is to preserve the leading directory `opensearch-<version+qualifier>` in zip
+            rootlen = len(self.archive_path)
+            leadingdirlen = len(os.path.basename(self.archive_path))
+            noleadingdirlen = rootlen - leadingdirlen
             for base, _, files in os.walk(self.archive_path):
                 for file in files:
                     fn = os.path.join(base, file)
-                    zip.write(fn, fn[rootlen:])
+                    zip.write(fn, fn[noleadingdirlen:])
 
 
 class DistRpm(Dist):

--- a/tests/tests_assemble_workflow/test_bundle_opensearch.py
+++ b/tests/tests_assemble_workflow/test_bundle_opensearch.py
@@ -251,5 +251,5 @@ class TestBundleOpenSearch(unittest.TestCase):
             with patch("shutil.copyfile") as mock_copyfile:
                 bundle.package(os.path.dirname(__file__))
                 mock_zipfile_open.assert_called_with("opensearch.zip", "w", zipfile.ZIP_DEFLATED)
-                mock_zipfile_write.assert_called_with(os.path.join(bundle.tmp_dir.name, "opensearch-1.3.0", "opensearch.txt"), "opensearch-1.3.0/opensearch.txt")
+                mock_zipfile_write.assert_called_with(os.path.join(bundle.tmp_dir.name, "opensearch-1.3.0", "opensearch.txt"), os.path.join("opensearch-1.3.0", "opensearch.txt"))
                 self.assertEqual(mock_copyfile.call_count, 1)

--- a/tests/tests_assemble_workflow/test_bundle_opensearch.py
+++ b/tests/tests_assemble_workflow/test_bundle_opensearch.py
@@ -251,5 +251,5 @@ class TestBundleOpenSearch(unittest.TestCase):
             with patch("shutil.copyfile") as mock_copyfile:
                 bundle.package(os.path.dirname(__file__))
                 mock_zipfile_open.assert_called_with("opensearch.zip", "w", zipfile.ZIP_DEFLATED)
-                mock_zipfile_write.assert_called_with(os.path.join(bundle.tmp_dir.name, "opensearch-1.3.0", "opensearch.txt"), "opensearch.txt")
+                mock_zipfile_write.assert_called_with(os.path.join(bundle.tmp_dir.name, "opensearch-1.3.0", "opensearch.txt"), "opensearch-1.3.0/opensearch.txt")
                 self.assertEqual(mock_copyfile.call_count, 1)

